File: app/webapp/src/lib/dtoValidation.ts
export const DTO_LIMITS = {
  postTitleMax: 128,
  postContentMax: 1024,
  commentContentMax: 512,
  userAccountMax: 128,
  userPasswordMin: 6,
  userPasswordMax: 128,
  userNameMax: 128,
  userAvatarMax: 128,
  userEmailMax: 128,
  userPhoneMax: 32,
  userInfoMax: 1024,
} as const;

export function requiredMax(value: string, max: number, emptyMessage: string, maxMessage: string) {
  const text = value.trim();
  if (!text) return emptyMessage;
  if (text.length > max) return maxMessage;
  return null;
}

export function optionalMax(value: string | null | undefined, max: number, maxMessage: string) {
  if ((value ?? "").trim().length > max) return maxMessage;
  return null;
}

export function passwordLength(value: string, emptyMessage: string, minMessage: string, maxMessage: string) {
  if (!value) return emptyMessage;
  if (value.length < DTO_LIMITS.userPasswordMin) return minMessage;
  if (value.length > DTO_LIMITS.userPasswordMax) return maxMessage;
  return null;
}

File: app/mobile_app/lib/utils/dto_validation.dart
class DtoLimits {
  const DtoLimits._();

  static const postTitleMax = 128;
  static const postContentMax = 1024;
  static const commentContentMax = 512;
  static const userAccountMax = 128;
  static const userPasswordMin = 6;
  static const userPasswordMax = 128;
  static const userNameMax = 128;
  static const userAvatarMax = 128;
  static const userEmailMax = 128;
  static const userPhoneMax = 32;
  static const userInfoMax = 1024;
}

String? requiredMaxValidator(
  String? value, {
  required String emptyMessage,
  required int max,
  required String maxMessage,
}) {
  final text = value?.trim() ?? '';
  if (text.isEmpty) return emptyMessage;
  if (text.length > max) return maxMessage;
  return null;
}

String? optionalMaxValidator(
  String? value, {
  required int max,
  required String maxMessage,
}) {
  final text = value?.trim() ?? '';
  if (text.length > max) return maxMessage;
  return null;
}

String? passwordLengthValidator(
  String? value, {
  required String emptyMessage,
  required String minMessage,
  required String maxMessage,
}) {
  final text = value ?? '';
  if (text.isEmpty) return emptyMessage;
  if (text.length < DtoLimits.userPasswordMin) return minMessage;
  if (text.length > DtoLimits.userPasswordMax) return maxMessage;
  return null;
}
